vermouth/ffinput.py

```diff
@@ -51,196 +55,275 @@
+        'impropers': 4,
+        'constraints': 2,
+        'virtual_sites2': 3,
+        'pairs': 2,
+    }
 
-    Returns
-    -------
-    str
-    """
-    start = None
-    while True:  # stops when start < 0
-        start = line.find('$', start)
-        if start < 0:
-            break
-        for end, char in enumerate(line[start + 1:], start=start + 1):
-            if char in ' \t\n{}$"':
-                break
-        else: # no break
-            end += 1
-        macro_name = line[start + 1:end]
-        macro_value = macros[macro_name]
-        line = line[:start] + macro_value + line[end:]
-        end = start + len(macro_value)
-    return line
+    def __init__(self, force_field):
+        super().__init__()
+        self.force_field = force_field
+        self.current_block = None
+        self.current_link = None
+        self.current_modification = None
+        self.blocks = collections.OrderedDict()
+        self.links = []
+        self.modifications = []
+
+        self.header_actions = {
+            ('moleculetype', ): self._new_block,
+            ('link', ): self._new_link,
+            ('modification', ): self._new_modification,
+        }
+
+    def parse_header(self, line, lineno=0):
+        """
+        Parses a section header with line number `lineno`. Sets :attr:`section`
+        when applicable. Does not check whether `line` is a valid section
+        header.
+
+        Parameters
+        ----------
+        line: str
+        lineno: str
+
+        Returns
+        -------
+        object
+            The result of calling :meth:`finalize_section`, which is called
+            if a section ends.
+
+        Raises
+        ------
+        KeyError
+            If the section header is unknown.
+        """
+        prev_section = self.section
+
+        ended = []
+        section = self.section + [line.strip('[ ]').casefold()]
+        if tuple(section[-1:]) in self.METH_DICT:
+            self.section = section[-1:]
+        else:
+            wildcard_section = tuple(section[:-1] + ['*'])
+            while (tuple(section) not in self.METH_DICT
+                   and wildcard_section not in self.METH_DICT
+                   and len(section) > 1):
+                ended.append(section.pop(-2))  # [a, b, c, d] -> [a, b, d]
+                wildcard_section = tuple(section[:-1] + ['*'])
+            self.section = section
+
+        result = None
+        if prev_section:
+            result = self.finalize_section(prev_section, ended)
+
+        action = self.header_actions.get(tuple(self.section))
+        if action is not None:
+            action()
+
+        return result
+
+    def parse_section(self, line, lineno):
+        """
+        Parse `line` with line number `lineno` by looking up the section in
+        :attr:`METH_DICT` and calling that method.
+
+        On the contrary to the SectionLineParser, we can have wildcard sections.
+        Wildcard sections cannot have children sections.
```

Comment:
@pckroon This parser doesn't do these type of .rtp style sections, which poses three new questions in my opinion:

1) why do we need a wild card parsing element anyways? Is it just because we want avoid writing a method for all bonds angles dihedrals generic interaction? In this case isn't this rather dangerous, because we read any generic interaction without actually knowing, if they are correct. For example typos are directly parsed through. 

2) Why is the rtp parser not a subclass of one of the generic parser classes (i.e. SecionLineParser)?

3) Why do some parsers live in the gmx folder and others on the main folder level of vermouth? Wouldn't it make more sense to have all parsers in one place, all writing utilities in one place. The question if they are GMX format or NAMD is in my opinion secondary.

Reply:
I agree with all your suggestions here. Let me get a bit more in depth.

1) For the ITP and FF parsers we don't. I strongly agree with the typo guarding by spelling out all allowed sections.
2) Because that didn't exist when the rtp parser was written.
3) I don't think it /really/ matters.

Reply:
@jbarnoud what's your take on issue 1?

Reply:
Issue 1 mostly came out of frustration from the first martinize: if gromacs comes out with a novel section I did not want martinize 2 to lag behind. Though, it happens rarely enough; typos are much more likely and the complexity the wildcard adds is probably not worth it.

Reply:
@pckroon shall we remove the wildcard section then?

Reply:
yes.